angular/src/components/login.component.ts

```diff
@@ -33,10 +34,14 @@ export class LoginComponent implements OnInit {
 
     masterPassword: string = '';
     showPassword: boolean = false;
+    captchaSiteKey: string = null;
+    captchaToken: string = null;
+    captcha: Captcha;
     formPromise: Promise<AuthResult>;
     onSuccessfulLogin: () => Promise<any>;
     onSuccessfulLoginNavigate: () => Promise<any>;
     onSuccessfulLoginTwoFactorNavigate: () => Promise<any>;
+    onCaptchaRequired: () => Promise<any>;
```

Comment:
Any particular reason this is handled as a callback? If not would it be appropriate to change `LoginComponent` to be an abstract class forcing this method to be implemented?

Reply:
It's because it was only needed for Desktop, where the login component width is too small to show the full captcha, so the `OnCaptchaRequired` is used to expand the `container` div to an appropriate width

Reply:
In that case, we could perhaps extend submit, and check if `captchaSiteKey` is set?